### PR TITLE
[staking] refactor to split staking abi into separate version

### PR DESCRIPTION
**Pull request by envestcc**
# Description

This update splits the staking web3 API into different versions, making it easier for future maintenance and updates of the interfaces.

Fixes #(issue)

## Type of change
Please delete options that are not relevant.
- [x] Code refactor or improvement

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: action/protocol/staking/ethabi/common/bucketsbycandidate.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+// EncodeToEth encode proto to eth
+func (r *BucketsByCandidateStateContext) EncodeToEth(resp *iotexapi.ReadStateResponse) (string, error) {
+	var result iotextypes.VoteBucketList
+	if err := proto.Unmarshal(resp.Data, &result); err != nil {
+		return "", err
+	}
+
+	return EncodeVoteBucketListToEth(r.Method.Outputs, &result)
```
Comment (dustinxie) [inline action/protocol/staking/ethabi/common/bucketsbycandidate.go:75]: looks like this code is the same for all the new contexts you added. Can move to `protocol.BaseStateContext` and become a common method?

File: action/protocol/staking/ethabi/stake_base.go [hunk 2]
```diff
@@ -1,203 +1,25 @@
...
-		return newCompositeBucketsByVoterStateContext(data[4:])
-	case hex.EncodeToString(_compositeBucketsCountMethod.ID):
-		return newCompositeBucketsCountStateContext()
-	case hex.EncodeToString(_compositeTotalStakingAmountMethod.ID):
-		return newCompositeTotalStakingAmountContext()
-	case hex.EncodeToString(_contractBucketTypesMethod.ID):
-		return newContractBucketTypesStateContext(data[4:])
-	default:
-		return nil, errInvalidCallSig
+	methodSig, err := v1.BuildReadStateRequest(data)
```
Comment (dustinxie) [inline action/protocol/staking/ethabi/stake_base.go:16]: why v1 takes precedence over v2?

File: action/protocol/staking/ethabi/stake_base.go [hunk 3]
```diff
@@ -1,203 +1,25 @@
...
-		return "", err
-	}
-	return hex.EncodeToString(data), nil
-}
-
 // BuildReadStateRequest decode eth_call data to StateContext
 func BuildReadStateRequest(data []byte) (protocol.StateContext, error) {
 	if len(data) < 4 {
-		return nil, errInvalidCallData
+		return nil, stakingComm.ErrInvalidCallData
```
Comment (dustinxie) [inline action/protocol/staking/ethabi/stake_base.go:13]: nit:
this is checked in v1 and v2's func, no need to check here?

File: action/protocol/staking/ethabi/common/bucketsbycandidate.go (hunk 1 above)
Comment (envestcc) [reply action/protocol/staking/ethabi/common/bucketsbycandidate.go:75]: It's acceptable to define `EncodeVoteBucketListToEth` same pkg with  `BucketEth`, and moving will cause cycle dependency problem

File: action/protocol/staking/ethabi/stake_base.go (hunk 2 above)
Comment (envestcc) [reply action/protocol/staking/ethabi/stake_base.go:16]: In principle, to maintain forward compatibility, each version should have completely new interfaces without interface overlapping.

File: action/protocol/staking/ethabi/stake_base.go (hunk 3 above)
Comment (envestcc) [reply action/protocol/staking/ethabi/stake_base.go:13]: okay

**Review by dustinxie — APPROVED** [review]

**Review by CoderZhi — APPROVED** [review]

**Pull request merged**